executor/common_usb.h

```diff
@@ -0,0 +1,495 @@
+};
+
+struct usb_fuzzer_event {
+	uint32 type;
+	uint32 length;
+	char data[0];
+};
+
+struct usb_fuzzer_init {
+	uint64 speed;
+	const char* driver_name;
+	const char* device_name;
+};
+
+struct usb_fuzzer_ep_io {
+	uint16 ep;
+	uint16 flags;
+	uint32 length;
+	char data[0];
+};
+
+#define USB_FUZZER_IOCTL_INIT _IOW('U', 0, struct usb_fuzzer_init)
+#define USB_FUZZER_IOCTL_RUN _IO('U', 1)
+#define USB_FUZZER_IOCTL_EP0_READ _IOWR('U', 2, struct usb_fuzzer_event)
+#define USB_FUZZER_IOCTL_EP0_WRITE _IOW('U', 3, struct usb_fuzzer_ep_io)
+#define USB_FUZZER_IOCTL_EP_ENABLE _IOW('U', 4, struct usb_endpoint_descriptor)
+#define USB_FUZZER_IOCTL_EP_WRITE _IOW('U', 6, struct usb_fuzzer_ep_io)
+#define USB_FUZZER_IOCTL_CONFIGURE _IO('U', 8)
+#define USB_FUZZER_IOCTL_VBUS_DRAW _IOW('U', 9, uint32)
+
+int usb_fuzzer_open()
+{
+	return open("/sys/kernel/debug/usb-fuzzer", O_RDWR);
+}
+
+int usb_fuzzer_init(int fd, uint32 speed, const char* driver, const char* device)
+{
+	struct usb_fuzzer_init arg;
+	arg.speed = speed;
+	arg.driver_name = driver;
+	arg.device_name = device;
+	return ioctl(fd, USB_FUZZER_IOCTL_INIT, &arg);
+}
+
+int usb_fuzzer_run(int fd)
+{
+	return ioctl(fd, USB_FUZZER_IOCTL_RUN, 0);
+}
+
+int usb_fuzzer_ep0_read(int fd, struct usb_fuzzer_event* event)
+{
+	return ioctl(fd, USB_FUZZER_IOCTL_EP0_READ, event);
+}
+
+int usb_fuzzer_ep0_write(int fd, struct usb_fuzzer_ep_io* io)
+{
+	return ioctl(fd, USB_FUZZER_IOCTL_EP0_WRITE, io);
+}
+
+int usb_fuzzer_ep_write(int fd, struct usb_fuzzer_ep_io* io)
+{
+	return ioctl(fd, USB_FUZZER_IOCTL_EP_WRITE, io);
+}
+
+int usb_fuzzer_ep_enable(int fd, struct usb_endpoint_descriptor* desc)
+{
+	return ioctl(fd, USB_FUZZER_IOCTL_EP_ENABLE, desc);
+}
+
+int usb_fuzzer_configure(int fd)
+{
+	return ioctl(fd, USB_FUZZER_IOCTL_CONFIGURE, 0);
+}
+
+int usb_fuzzer_vbus_draw(int fd, uint32 power)
+{
+	return ioctl(fd, USB_FUZZER_IOCTL_VBUS_DRAW, power);
+}
+
+#define USB_MAX_PACKET_SIZE 1024
+
+struct usb_fuzzer_control_event {
+	struct usb_fuzzer_event inner;
+	struct usb_ctrlrequest ctrl;
+};
+
+struct usb_fuzzer_ep_io_data {
+	struct usb_fuzzer_ep_io inner;
+	char data[USB_MAX_PACKET_SIZE];
+};
+
+struct vusb_connect_string_descriptor {
+	uint32 len;
+	char* str;
+} __attribute__((packed));
+
+struct vusb_connect_descriptors {
+	uint32 qual_len;
+	char* qual;
+	uint32 bos_len;
+	char* bos;
+	uint32 strs_len;
+	struct vusb_connect_string_descriptor strs[0];
+} __attribute__((packed));
+
+static uintptr_t syz_usb_connect(uintptr_t a0, uintptr_t a1, uintptr_t a2, uintptr_t a3, uintptr_t a4)
+{
+	int64_t speed = a0;
+	int64_t dev_len = a1;
+	char* dev = (char*)a2;
+	char* dev_id = (char*)a3;
+	struct vusb_connect_descriptors* conn_descs = (struct vusb_connect_descriptors*)a4;
+
+	debug("syz_usb_connect: dev: %p, dev_id: %p\n", dev, dev_id);
+	if (!dev)
+		return -1;
+
+	debug("syz_usb_connect: device data:\n");
+	debug_dump_data(dev, dev_len);
+
+	if (dev_id)
+		patch_usb_descriptor(dev, dev_id);
```

Comment:
Let's make generateUsbDeviceID directly fill in fields in dev. Then we don't need usb_device_id and patch_usb_descriptor.

Reply:
Done.